mempool/iterators.go

```diff
@@ -116,45 +116,48 @@ func NewBlockingIterator(mem *CListMempool) Iterator {
 func (iter *BlockingIterator) WaitNextCh() <-chan Entry {
 	ch := make(chan Entry)
 	go func() {
+		// Pick a lane containing the next entry to access.
+		lane, ok := iter.pickLane()
+		if !ok {
+			// There are no transactions to take from any lane. Wait until a new
+			// transaction is added to the mempool.
+			<-iter.mp.addedTxCh()
+			lane = iter.mp.latestTxLane()
+		}
+
 		// Add the next entry to the channel if not nil.
-		lane := iter.PickLane()
-		if entry := iter.Next(lane); entry != nil {
+		if entry := iter.next(lane); entry != nil {
 			ch <- entry.Value.(Entry)
 		}
-		// Unblock the receiver (it may receive nil).
+
+		// Unblock receiver if no entry was sent (it may receive nil).
 		close(ch)
 	}()
 	return ch
 }
 
-// PickLane returns a _valid_ lane on which to iterate, according to the WRR
-// algorithm. A lane is valid if it is not empty or it is not over-consumed,
-// meaning that the number of accessed entries in the lane has not yet reached
-// its priority value in the current WRR iteration. It will block until a
-// transaction is available in any lane.
-func (iter *BlockingIterator) PickLane() types.Lane {
+// pickLane returns a _valid_ lane containing the next transaction to access
+// according to the WRR algorithm. A lane is valid if it is not empty or it is
+// not over-consumed, meaning that the number of accessed entries in the lane
+// has not yet reached its priority value in the current WRR iteration. It
+// returns false if all lanes are empty or don't have transactions that have not
+// yet been accessed.
+func (iter *BlockingIterator) pickLane() (types.Lane, bool) {
 	// Start from the last accessed lane.
 	lane := iter.sortedLanes[iter.laneIndex]
 
-	iter.mp.addTxChMtx.RLock()
-	defer iter.mp.addTxChMtx.RUnlock()
-
```

Comment:
I see `pickLane` no longer takes a lock on the mempool. I think that's why you turned `lastTxSeq` into a `sync.Map`.

I know this is the "most performant", but I'd still consider taking a read lock on the mempool (i.e., `txsMtx`). Why? Because not holding the lock means we're searching the mempool while it is potentially being modified. For instance (and we have considered this example a few weeks ago):
* mempool empty
* we call `pickLane()`
* while the `for` loop in line 145 is about to return in line 160
  * a new tx arrives
  * we add the tx to a lane the `for` loop has already visited
  * we close `addTxCh`, and create a new one
* `pickLane()` returns in line 160
* `WaitNextCh` blocks on the newly created `addTxCh`
* We end up in a situation where a tx is in the mempool, available for this iterator, but we are blocked on `addTxCh`

The example can be made worse replacing "a tx" by "a number of tx".

I know this is not likely to cause big problems in practice, but those quirks are never fully understood until they bite hard (days of troubleshooting).

So, if taking a read lock here is not impacting performance, I'd prefer that way, as the whole concurrency model here is much easier to understand.

Reply:
100% agree. Since we'll soon be running a "dry-run" QA test with 200 nodes, we can be seeing there if taking `txsMtx` (read lock) is impacting performance w.r.t. `v1.x`. If we indeed see performance suffers, we can have a patch ready with the version where `pickLane` doesn't take the lock to "swap in". WDYT?